brave/brave-core
# Implement NTP Custom homepage ("super referral")

Resolves https://github.com/brave/brave-browser/issues/8218

`NTPBackgroundImagesService`manages background provider(SI or SR).
By default, it registers SI component and SR component if it's super referral install.

It's global service because SI or SR component is browser-wide component and it manages their data itself. Which data should be used for webui or android NTP will be determined by `ViewCounterService`. `ViewCounterService` will provide proper data (SI or SR) to client (like webui) by checking current profile's preference.

`NTPBackgroundImagesSource` handles bg data request. It can handle SI and SR data simultaneously.
SI data's url starts with `chrome://branded-wallpaper/sponsored-images/...` and SR data is something like  `chrome://branded-wallpaper/super-referral/logo.png`.
I think we can expand this in the future. For example, `chrome://branded-wallpaper/default/wallpaper.jpg` for default background.

To make SR data available quickly, SR data is cached. With this caching, client can get SR data
before component registration is ready. It's important because SR's NTP always should display its image.

In this PR, `kNewTabPageSuperReferralThemesOption` pref is added.
Client should use it as a switch for turn it on/off SR component in runtime in super referral installs.
If it's turned off, SI or default bg will be displayed. This can be changed in settings page.

Regarding to flags, existing `New Tab Page Branded Wallpapers` flag will act as a top level switch.
If it is off, `SI` and `SR` are all disabled.
If `New Tab Page Branded Wallpapers` feature switch is on, `SI` component is enabled by default.
New `Enable Brave Custom Homepage` is introduced for `SR` component. If this is on, `SR` component will be installed when current install comes from super referrer.

To test locally, we can pass different local test data by using `ntp-sponsored-images-data-path` and `ntp-super-referral-data-path`. Also can use demo components. It becomes available when demo flag is on.

### Settings page with Super referral ###
<img width="791" alt="Screen Shot 2020-03-26 at 2 25 00 PM" src="https://user-images.githubusercontent.com/6786187/77613464-1c4b3d00-6f6e-11ea-91b4-573dfa8b4f8b.png">

![image](https://user-images.githubusercontent.com/6786187/78313902-3fd83e00-7593-11ea-9dae-2cfa117d7da0.png)

### Settings page with non Super referral ###
<img width="804" alt="Screen Shot 2020-03-26 at 2 26 54 PM" src="https://user-images.githubusercontent.com/6786187/77613523-3e44bf80-6f6e-11ea-9a50-fc48a3096b3d.png">

### NTP Pages with Super referral ###
![image](https://user-images.githubusercontent.com/6786187/78313831-0bfd1880-7593-11ea-9d95-dd14ec4018d6.png)



## Submitter Checklist:

- Verified that these changes build without errors on
- Verified that these changes pass automated tests (unit, browser, security tests) on

## Test Plan:
**Simply test with demo data** => SR component test is not supported by Demo flag.
~~1. Launch browser with fresh profile
2. Turn on `New Tab Page Demo Branded Wallpaper` in brave://flags and relaunch
3. Open brave://components and check SI and SR component is registered (`NTP Sponsored Images (DEMO)` and `NTP Super Referral (Technikke)`)
4. Open NTP and check SR is loaded. If top sites are visible, it means SR component is loaded
*Note:* With demo data, theme settings are not properly updated. To see the themes option change, please follow below step.~~

**Check fresh install with super referrer**
1. Create empty profile dir
2. Create promoCode text file in root profile dir and that file has `TECHNIK` string
3. Launch Browser and check `NTP Super Referrer (Technikke)` component is registered in brave://components
4. Load new tab and topsites are visible for every NTP loading
5. Open settings page and check themes option is changed in appearance settings
6. Click theme option and check new themes sub page.
7. Check super referral bg is not displayed after selecting `Brave default images` option.

### Simulate campaign ends(continued test from above)
8. Launch browser and turn custom home page flag off
9. Load NTP and top sites from SR still visible well

**Check fresh install with non referrer**
1. Launch browser with clean profile
2. Check NTP SI component is only registered

### Two scenarios that check this change doesn't affect existing users
***Scenario 1 (User has default code or non super referrer code)***
1. Create user dir with other builds
2. Launch browser with this PR with created user data at step 1
3. Check NTP Sponsored Images component is only registered

***Scenario 2 (User has empty code because of 90 days passed)***
1. Copy old user data dir that used more than 90 days
2. Start browser with that copied user dir
3. Check NTP Sponsored Images component is only registered

## Reviewer Checklist:


## After-merge Checklist:

  changes has landed on.
